### PR TITLE
Create tests and optimize rendering of AllThreadsCallstackBar

### PR DESCRIPTION
**Pull request by karupayun**
In this PR we are optimizing the rendering of AllThreadsCallstackBar.

In the previous PR (https://github.com/google/orbit/pull/4271), we skip going through and rendering events from the same pixel in a particular thread_id but we were still going once per pixel per thread_id. In this PR we are still iterating through pixels but stopping looking for new callstacks when we have already found some on the current_pixel.

For getting the pixel number I created a method in FastRenderingUtils class, but I also used in CallstackThreadBar since is way more precise. Ideally we should fix these functions (http://b/242973688), but I won't work on that for now. It also makes sense to use GetPixelNumber when we want to render vertical lines.

I found this issue when creating tests for CallstackData (there were no tests before for this method but I should have created them anyway), so I'm including the tests here as well.

Rendering all_thread Track for a 5 min-capture takes:
- Before (https://github.com/google/orbit/pull/4271): >150ms
- After (https://github.com/google/orbit/pull/4271): ~13ms
- Now: ~0.950ms

Rendering callstacks for all Tracks takes 8.6% instead of 11.9% (and before https://github.com/google/orbit/pull/4271 - 66%) . I think that this is enough for our goal.

Test: Load a capture, rendering looks fine. Run Unit tests.
Bugs: http://b/252998254 (GetPixelNumber) & http://b/253414641 (Callstack optimization itself)

### CODE REVIEW

File: src/ClientData/include/ClientData/CallstackData.h
Comment (dpallotti) [inline src/ClientData/include/ClientData/CallstackData.h:87]: `constexpr`

File: src/ClientData/include/ClientData/CallstackData.h
Comment (dpallotti) [inline src/ClientData/include/ClientData/CallstackData.h:91]: I was confused because from the name it sounds like this is a tid, not a callstack (and `auto` doesn't help). I would use `next_callstack_of_tid`.

File: src/ClientData/include/ClientData/CallstackData.h
Comment (dpallotti) [inline src/ClientData/include/ClientData/CallstackData.h:96]:
```suggestion
          // If this callstack will be drawn in the current_pixel, we don't need to search for more
          // of them. Otherwise, there could be a callstack in another thread_id that will be drawn
          // before, so we need to keep looking.
```

File: src/ClientData/CallstackDataTest.cpp
Comment (dpallotti) [inline src/ClientData/CallstackDataTest.cpp:209]: I find it very strange to use a name like "kCloneAddress" when this test is just about computing timestamps and pixels and has nothing to do with the actual content of the callstacks.

File: src/ClientData/CallstackDataTest.cpp
Comment (dpallotti) [inline src/ClientData/CallstackDataTest.cpp:213]: I don't think there's any point declaring `cs#_outer`, `cs#_inner` as you are using them once and don't care about the values of the frames in this test. Just inline them in the construction of the `CallstackInfo`s.

File: src/ClientData/CallstackDataTest.cpp
Comment (dpallotti) [inline src/ClientData/CallstackDataTest.cpp:220]: Do you even need two distinct `CallstackInfo`s for this test? I feel like a single `CallstackInfo` with no frames at all would serve the same purpose.

Comment (dpallotti) [reply src/ClientData/CallstackDataTest.cpp:220]: In that case the other two comments I left on this test are irrelevant.

File: src/ClientData/include/ClientData/CallstackData.h
Comment (akopich) [inline src/ClientData/include/ClientData/CallstackData.h:89]: The name of variable and the loop right after its declaration make me wonder if its value changes. And it seems like it doesn't. I suggest `const` here.

File: src/ClientData/include/ClientData/CallstackData.h
Comment (akopich) [inline src/ClientData/include/ClientData/CallstackData.h:88]: Why can't we just return `optional<CallstackEvent>` instead of instroducing yet another special constant?

File: src/ClientData/include/ClientData/CallstackData.h
Comment (akopich) [reply src/ClientData/include/ClientData/CallstackData.h:89]: ... and as we const one variable, also const all the other constable variables.

File: src/ClientData/include/ClientData/CallstackData.h
Comment (akopich) [inline src/ClientData/include/ClientData/CallstackData.h:107]: why calling it `next_event_it` if it isn't an iterator?

File: src/OrbitGl/CallstackThreadBar.cpp
Comment (akopich) [inline src/OrbitGl/CallstackThreadBar.cpp:121]: Irrelevant to this PR, but IMO we either const all of the constable local variables, or none.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:206]: Just wondering. Why are some literal-initialized constants const and some constexpr?

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:211]: I'd prefer to put these right above the `TEST...`, so they would get static storage duration. But TBH it doesn't matter here.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:227]: A suggestion. You need to construct 4 CallstackEvents. So another option is to hard-code two lists of 4 elements -- `kTids` and `kTimes`. Then you can `absl::c_transfrom` the two lists and yield a collection of CallstackEvent.  And finally, populate `callstack_data` by a simple loop over the collection. This would cut code duplication. And all of these can be global variables.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:232]: 0 and 1000 could be variables.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:234]: Irrelevant to the PR, but `using ::testing::Pointwise` could make the code pretttier.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:254]: Could we de-duplicate this? E.g. have a function that would take the arguments for `ForEachCallstackEventOfTidInTimeRangeDiscretized` and the expected `std::vector<CallstackEvent>` ?

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:290]: Is this a copy-paste from the code above? Let's make these variables global.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:320]: Just like in the test above, I'd add a function. But it would have to take a matcher, not just a `vector<CallstackEvent>`.

**Review by akopich — COMMENTED** [review]
Nitpicking mostly.

File: src/ClientData/CallstackDataTest.cpp
Comment (nputikhin) [inline src/ClientData/CallstackDataTest.cpp:203]: I suggest to split these tests into small focused cases (see http://go/tott/520)

File: src/ClientData/CallstackDataTest.cpp
Comment (nputikhin) [reply src/ClientData/CallstackDataTest.cpp:234]: Agree: It is generally recommended to use `using` declaration for matchers (http://go/c-readability-advice#using-the-testing-namespace-with-expect-that)

File: src/ClientData/CallstackDataTest.cpp
Comment (nputikhin) [reply src/ClientData/CallstackDataTest.cpp:232]: I think this won't be needed with small focused test cases - these variables are fairly trivial and just set a time range

File: src/ClientData/include/ClientData/FastRenderingUtils.h
Comment (nputikhin) [inline src/ClientData/include/ClientData/FastRenderingUtils.h:17]: Please add a function comment (http://go/cstyle#function_declarations)

File: src/ClientData/FastRenderingUtilsTest.cpp
Comment (nputikhin) [inline src/ClientData/FastRenderingUtilsTest.cpp:16]: Consider parametrizing the test (https://go/parameterized-testing)

File: src/ClientData/include/ClientData/CallstackData.h
Comment (nputikhin) [reply src/ClientData/include/ClientData/CallstackData.h:88]: This should be be either `std::optional` (if owning) or a possibly-null pointer (if non-owning). See http://go/totw/171

File: src/ClientData/include/ClientData/CallstackData.h
Comment (nputikhin) [inline src/ClientData/include/ClientData/CallstackData.h:92]: Propose to simplify the flow by splitting the nested ifs into several conditions like:
- if the callstack candidate is end, continue
- if the callstack candidate is within the pixel, return it
- if the callstack candidate is earlier than the next callstack, set the next callstack to it

This would make the flow a bit easier to follow

File: src/ClientData/include/ClientData/CallstackData.h
Comment (karupayun) [reply src/ClientData/include/ClientData/CallstackData.h:88]: This constants are not used, it's just to fill the callstack with some ID. Initially I implemented as an optional but the condition was a bit harder to read. But I agree that using max_timestamp here is strange, will convert it back to optional.

File: src/OrbitGl/CallstackThreadBar.cpp
Comment (karupayun) [reply src/OrbitGl/CallstackThreadBar.cpp:121]: I agree with what you said although I might prefer in some cases to have some const statements.

I will const all the lines I changed but I won't add const in previous existent variables.

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp:206]: It was like that before, I've  just copy pasted the previous existent tests.

Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: I made some changes, PTAL.

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: Done

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: I made a big refactor, PTAL.

File: src/ClientData/FastRenderingUtilsTest.cpp
Comment (karupayun) [reply src/ClientData/FastRenderingUtilsTest.cpp]: Done. Also done for the other method GetNextPixelBoundaryTimeNs.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [reply src/ClientData/CallstackDataTest.cpp]: This is what I had in mind:

```
const std::vector<uint32_t> kTids = {kTid, kTid, kAnotherTid, kTid};
const std::vector<uint64_t> kTimestamps = {142, 242, 342, 442};

const std::vector<CallstackEvent> kAllEvents = [] {
  std::vector<CallstackEvent> events;
  absl::c_transform(kTids, kTimestamps, std::back_inserter(events),
                    [](uint32_t tid, uint64_t timestamp) {
                      return CallstackEvent{timestamp, kCallstackId1, tid};
                    });
  return events;
}();
```

This way way we don't have to introduce a named function, call it multiple times, introduce the `TidAndTimestamp` struct. 

OTOH, feel free to keep your solution. Just rename the vector `kTimestampsAndTidInForEachCallstackTests`. I think, `kTidAndTimestampInForEachCallstackTest` suits it better.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [reply src/ClientData/CallstackDataTest.cpp]: The variables still seem to be cope-pasted, but you've given a thumb up. Have you missed a commit?

Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: I didn't touch the previous existent tests too much and I've just modified the new cases where the callstack_id doesn't matter.

TBH, I didn't take a real look to the previous existent tests. I was waiting for approval of this PR, and I can try to do something similar for the other cases, although the events differ (no fully repeated code). I can make it in a following PR.

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: The TidAndTimestamp struct was created just to assure that the 2 vectors have the same size. But TBH I don't have a strong opinion about that, I can divided between 2 different vectors since it's a test.

**Review by akopich — APPROVED** [review]
Thanks for addressing my concerns.
I've left some more comments :) 
LGTM otherwise! Feel free to re-request the review if you would like me to take another look.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [reply src/ClientData/CallstackDataTest.cpp]: OK.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [reply src/ClientData/CallstackDataTest.cpp]: You can make sure that the length is the same by using `std::array` instead of `std::vector` for `kTids` and `kTimestamps`.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [reply src/ClientData/CallstackDataTest.cpp:223]: And I'd rather call use `indices` instead of `desired_positions`.

File: src/ClientData/CallstackDataTest.cpp
Comment (akopich) [inline src/ClientData/CallstackDataTest.cpp:286]: Just a thought. 

I don't like this enum. One way to circumvent the need for it would be to introduce two lambdas

```
constexpr auto kExpectAll = [](const auto& actual, const auto& expected) {
  EXPECT_THAT(actual, Pointwise(CallstackEventEq(), expected));
};

constexpr auto kExpectAny = [](const auto& actual, const auto& expected) {
  EXPECT_EQ(actual.size(), 1);
  EXPECT_THAT(actual[0], AnyOfArray(expected));
};
```
And replace the field `ExpectType expect_type` in `FormattingForEachCallstackEventInTimeRangeDiscretizedTest` with `absl::FunctionRef<void(std::vector<CallstackEvent>, std::vector<CallstackEvent>)> expect;`. Then in the test you can just call 
```
test_case.expect(visited_callstack_list, GetSubset(kAllEvents, test_case.expected_event_ids));
```

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: Done. I mixed it up with insertion of events.

File: src/ClientData/CallstackDataTest.cpp
Comment (karupayun) [reply src/ClientData/CallstackDataTest.cpp]: Yes, sure. I will keep vectors as it is a test.

**Comment by karupayun** [comment]
> Thanks for addressing my concerns. I've left some more comments :) LGTM otherwise! Feel free to re-request the review if you would like me to take another look.

It's done. Let me know if you have more comments/want to take another look. Otherwise I will merge it after having another approval.

**Review by nputikhin — APPROVED** [review]

**Pull request merged**
